src/utils.jl

```diff
@@ -313,50 +202,104 @@ function split_dataset(data::Matrix; train_ratio::Float64 = 0.9, randomize::Bool
   return train_data,test_data
 end
 
+"""
+    _ischoi(M::LPDO)
 
+Check whether a given LPDO{MPO}  
+"""
+_ischoi(M::LPDO{MPO}) = (length(inds(M.X[1],"Site")) == 2 && hastags(M.X[1],"Purifier"))
 
-function ischoi(M::LPDO)
-  return (length(inds(M.X[1],"Site")) == 2 ? true : false)
-end
+_ischoi(M::MPO) = length(inds(M[1],"Site")) == 4 
+
+
+
+"""
+    _haschoitags(L::LPDO)
+
+    _haschoitags(M::Union{MPS,MPO})
+
+Check whether the TN has input/output Choi tags
+"""
+_haschoitags(L::LPDO) = (hastags(inds(L.X[1]),"Input") && hastags(inds(L.X[1]),"Output"))
+_haschoitags(M::Union{MPS,MPO}) = _haschoitags(LPDO(M)) 
 
-function ischoi(M::MPO)
-  return (length(inds(M[1],"Site")) == 4 ? true : false)
-  #return ( length(inds(M[1])) == 5 ? true : false)
-end
 
-function makeUnitary(L::LPDO{MPS})
-  ψ = L.X
-  U = MPO(ITensor[copy(ψ[j]) for j in 1:length(ψ)])
-  prime!(U,tags="Output")
-  removetags!(U, "Input")
-  removetags!(U, "Output")
-  return U
+"""
+    _choitags(U::Union{MPS, MPO})
+
+Assign the input/output tags defined for a Choi matrix to an MPO.
+
+  σ₁ -o- σ₁′       σ₁ⁱ -o- σ₁ᴼ   
+      |                 |
+  σ₂ -o- σ₂′  ⟶    σ₂ⁱ -o- σ₂ᴼ
+      |                 |
+  σ₃ -o- σ₃′       σ₃ⁱ -o- σ₃ᴼ
+                  
+"""
+function _choitags(L::LPDO{MPO})
+  _haschoitags(L) && return L
+  U = copy(L.X)
+  U = addtags(U, "Input", plev = 0, tags = "Qubit")
+  U = addtags(U, "Output", plev = 1, tags = "Qubit")
+  return LPDO(noprime(U))
 end
 
-function makeChoi(U0::MPO)
-  M = MPS(ITensor[copy(U0[j]) for j in 1:length(U0)])
-  addtags!(M, "Input", plev = 0, tags = "Qubit")
-  addtags!(M, "Output", plev = 1, tags = "Qubit")
-  noprime!(M)
-  return LPDO(M)
+_choitags(U::MPO) = _choitags(LPDO(U)).X
+
+"""
+    _mpotags(Ψ::MPS)
+
+Inverse of _choitags
+"""
+
+function _mpotags(L::LPDO)
+  !_haschoitags(L) && return L
+  U = copy(L.X) 
+  U = prime(U,tags="Output")
+  U = removetags(U, "Input")
+  U = removetags(U, "Output")
+  return LPDO(U)
 end
 
+_mpotags(M::Union{MPS,MPO}) = _mpotags(LPDO(M)).X
 
-function numberofqubits(gate::Tuple)
-  s = gate[2]
+
+"""
+    _unitaryMPO_to_choiMPS(L::LPDO{MPO})
+    _unitaryMPO_to_choiMPS(U::MPO)
+
+
+     MPO          MPS (vectorized MPO) 
+  σ₁ -o- σ₁′       o= (σ₁ⁱ,σ₁ᴼ)   
+      |            | 
+  σ₂ -o- σ₂′  ⟶    o= (σ₂ⁱ,σ₂′ᴼ)
+      |            | 
+  σ₃ -o- σ₃′       o= (σ₃ⁱ,σ₃′ᴼ)
+                  
+Transforms a unitary MPO into a Choi MPS with appropriate tags.
+"""
+#_unitaryMPO_to_choiMPS(L::LPDO{MPO}) = LPDO(convert(MPS, _choitags(L).X)) 
+_unitaryMPO_to_choiMPS(U::MPO) = convert(MPS, _choitags(U))
+_unitaryMPO_to_choiMPS(L::LPDO{MPO}) = LPDO(_unitaryMPO_to_choiMPS(L.X))
```

Comment:
Maybe this would better as:
```julia
unitary_mpo_to_choi_lpdo(U::MPO) = LPDO(unitary_mpo_to_choi_mps(U))
```
It seems best to stay consistent about not making `LPDO{MPO}` out of unitaries MPOs (maybe you have a use case for that).

Reply:
In fact, maybe this function isn't necessary, and instead you could just use `LPDO(unitary_mpo_to_choi_mps(U))` if you want an LPDO.